File: principlesJava/encap.java
//this defines  a public class called Person
public class Person {
    public String name;        // name is a  Public attribute meaning it can be accessed anywhere
    protected int age;         // age is a Protected attribute meaning it can be aceessed within the same package and by subclasses
    private String ssn;        // ssn is a Private attribute meaning it can be accessed within the Person class

    //  the Constructor initialises the name,age and ssn attributes when a new person object is created
    public Person(String name, int age) {
        this.name = name;
        this.age = age;
        this.ssn = "123-45-6789";
    }

    // Public method thats prints the name and age
    public void displayInfo() {
        System.out.println("Name: " + name + ", Age: " + age);// system.out is provided by java and it prints text to the console
    }                                                         //printIn ,this method of the system.out, prints the given text followed by a new line
                                                             //void is a method that performs an action but doesnt return any value

    // Private method that prints the ssn attributes and it can only be callled within the person class
    private void displaySSN() {
        System.out.println("SSN: " + ssn);
    }
}

// Main class
public class Main {
    public static void main(String[] args) {
        Person person = new Person("Alice", 30);

        // Accessing public attribute
        System.out.println(person.name);  // Output: Alice

        // Accessing protected attribute
        System.out.println(person.age);   // Output: 30

        // Accessing private attribute (will cause an error)
        // System.out.println(person.ssn); // Uncommenting this line will cause an error

        // Accessing private method (will cause an error)
        // person.displaySSN(); // Uncommenting this line will cause an error

        // Using public method
        person.displayInfo();  // Output: Name: Alice, Age: 30
    }
}

File: PrinciplesPython/encap.py
class Person:# defines a class Person
    def _init_(self, name, age):# a constructor is called for a newly formed  instance of a class  eg for class Person
        self.name = name        # Public attribute and assigned to the value passed to the constructor
        self._age = age         # Protected attribute(for internal use only)
        self.__ssn = "123-45-6789"  # Private attribute(cant be acessed directly outside the class)

    def display_info(self):
        print(f"Name: {self.name}, Age: {self._age}") #this method is public and it prints the name and age attributes that can be called from outside the class

    def __display_ssn(self):
        print(f"SSN: {self.__ssn}")#this method is private and it  prints the ssn and cannot be called directly from outside the class

# Creating an object
person = Person("Alice", 30)

# this line prints a name attribute of the object person and since name is public then it can be acessed directly
print(person.name)  # Output: Alice

# Accessing protected attribute
print(person._age)  # Output: 30

# Accessing private attribute (will raise an AttributeError)
# print(person.__ssn)  # Uncommenting this line will raise an error

# you  can acess private attributes via name mangling, a technique python uses to make private attributes harder to acess but incase u know the 
#the mangled name then u ca acess the private attributes
print(person.Person_ssn)  # Output: 123-45-6789
